Parallel end to end tests timing out

It appears that #1963 introduced some conditions where parallel operations on kopia will hang. This is causing some CI runs to fail tests as they timeout. I used a short timeout in this example, but one of my CI runs failed in a similar manner after ~500s

local output from commit before 1963:
```
root@cc40a4116d3d:/go/src/github.com/kopia/kopia# git checkout 17c74e63
root@cc40a4116d3d:/go/src/github.com/kopia/kopia# go test -tags testing -timeout 1m -count 10 \
    -run TestPlaceholderAndRealFails ./tests/end_to_end_test/...
ok      github.com/kopia/kopia/tests/end_to_end_test    0.558s
```

local output with 1963 merged:
```
root@cc40a4116d3d:/go/src/github.com/kopia/kopia# git checkout f5c64c84
root@cc40a4116d3d:/go/src/github.com/kopia/kopia# go test -tags testing -timeout 1m -count 10 \
    -run TestPlaceholderAndRealFails ./tests/end_to_end_test/...
panic: test timed out after 1m0s

goroutine 2376 [running]:
testing.(*M).startAlarm.func1()
  /usr/local/go/src/testing/testing.go:2029 +0x8c
created by time.goFunc
  /usr/local/go/src/time/sleep.go:176 +0x3c

goroutine 1 [chan receive]:
testing.tRunner.func1()
  /usr/local/go/src/testing/testing.go:1405 +0x458
testing.tRunner(0x4000396b60, 0x40002a7bc8)
  /usr/local/go/src/testing/testing.go:1445 +0x14c
testing.runTests(0x400055bf40?, {0x1f2bd80, 0x3c, 0x3c}, {0x2?, 0x3?, 0x1f3ea40?})
  /usr/local/go/src/testing/testing.go:1837 +0x3e8
testing.(*M).Run(0x400055bf40)
  /usr/local/go/src/testing/testing.go:1719 +0x510
github.com/kopia/kopia/internal/testutil.MyTestMain(0x1f3f920?, {0x40002a7e88, 0x1, 0xffa090?})  /go/src/github.com/kopia/kopia/internal/testutil/testutil.go:109 +0x40
github.com/kopia/kopia/tests/end_to_end_test_test.TestMain(0x400006fec8?)
  /go/src/github.com/kopia/kopia/tests/end_to_end_test/main_test.go:82 +0x7c
main.main()
  _testmain.go:167 +0x1ec

goroutine 50 [chan receive]:
github.com/golang/glog.(*loggingT).flushDaemon(0x400005c79e?)
  /go/pkg/mod/github.com/golang/glog@v1.0.0/glog.go:882 +0x60
created by github.com/golang/glog.init.0
  /go/pkg/mod/github.com/golang/glog@v1.0.0/glog.go:410 +0x204

goroutine 68 [select]:
go.opencensus.io/stats/view.(*worker).start(0x4000290080)
  /go/pkg/mod/go.opencensus.io@v0.23.0/stats/view/worker.go:276 +0x88
created by go.opencensus.io/stats/view.init.0
  /go/pkg/mod/go.opencensus.io@v0.23.0/stats/view/worker.go:34 +0xa8

goroutine 1913 [semacquire]:
sync.runtime_Semacquire(0x40001029c0?)
  /usr/local/go/src/runtime/sema.go:56 +0x2c
sync.(*WaitGroup).Wait(0x4000045160)
  /usr/local/go/src/sync/waitgroup.go:136 +0x88
github.com/kopia/kopia/tests/testenv.(*CLITest).Run(0x40000bbe80, 0x4000396d00, 0x1, {0x40003da090, 0x3, 0x4000396d00?})
  /go/src/github.com/kopia/kopia/tests/testenv/cli_test_env.go:220 +0x478
github.com/kopia/kopia/tests/testenv.(*CLITest).RunAndExpectFailure(0x164afa0?, 0x4000396d00, {0x40003da090, 0x3, 0x3})
  /go/src/github.com/kopia/kopia/tests/testenv/cli_test_env.go:145 +0x58
github.com/kopia/kopia/tests/end_to_end_test_test.TestPlaceholderAndRealFails(0x4000396d00)
  /go/src/github.com/kopia/kopia/tests/end_to_end_test/shallowrestore_test.go:489 +0x584
testing.tRunner(0x4000396d00, 0x14a8de8)
  /usr/local/go/src/testing/testing.go:1439 +0x110
created by testing.(*T).Run
  /usr/local/go/src/testing/testing.go:1486 +0x328

goroutine 2208 [select]:
io.(*pipe).read(0x40003de600, {0x4001338521, 0xadf, 0x88c54?})
  /usr/local/go/src/io/pipe.go:57 +0x84
io.(*PipeReader).Read(0x4001338373?, {0x4001338521?, 0x4000567df8?, 0x88dec?})
  /usr/local/go/src/io/pipe.go:136 +0x28
bufio.(*Scanner).Scan(0x4000567f40)
  /usr/local/go/src/bufio/scan.go:215 +0x89c
github.com/kopia/kopia/tests/testenv.(*CLITest).Run.func2()
  /go/src/github.com/kopia/kopia/tests/testenv/cli_test_env.go:211 +0xac
created by github.com/kopia/kopia/tests/testenv.(*CLITest).Run
  /go/src/github.com/kopia/kopia/tests/testenv/cli_test_env.go:207 +0x470

goroutine 2206 [semacquire]:
sync.runtime_Semacquire(0x400056fa40?)
	/usr/local/go/src/runtime/sema.go:56 +0x2c
sync.(*WaitGroup).Wait(0x4000fe4f20)
	/usr/local/go/src/sync/waitgroup.go:136 +0x88
github.com/kopia/kopia/internal/workshare.(*Pool).Close(0x4000fe4f00)
	/go/src/github.com/kopia/kopia/internal/workshare/workshare_pool.go:80 +0x38
github.com/kopia/kopia/snapshot/snapshotfs.(*Uploader).Upload(0x4000260b40, {0x164ef50?, 0x4001516000?}, {0x1654508?, 0x40002af4a0}, 0x400000eb58, {{0x40004e8a90, 0xc}, {0x40004e89fc, 0x4}, ...}, ...)
	/go/src/github.com/kopia/kopia/snapshot/snapshotfs/upload.go:1156 +0x654
github.com/kopia/kopia/cli.(*commandSnapshotCreate).snapshotSingleSource(0x40014e2888, {0x164ef50, 0x4001516000}, {0xffff527fa638, 0x40000df600}, 0x610001d27c?, {{0x40004e8a90, 0xc}, {0x40004e89fc, 0x4}, ...}, ...)
	/go/src/github.com/kopia/kopia/cli/command_snapshot_create.go:299 +0x4cc
github.com/kopia/kopia/cli.(*commandSnapshotCreate).run(0x40014e2888, {0x164ef50, 0x4001516000}, {0xffff527fa638?, 0x40000df600})
	/go/src/github.com/kopia/kopia/cli/command_snapshot_create.go:142 +0x5b8
github.com/kopia/kopia/cli.(*App).repositoryWriterAction.func1.1({0x164ef50?, 0x4001516000?}, {0xffff527fa638?, 0x40000df600?})
	/go/src/github.com/kopia/kopia/cli/app.go:418 +0x38
github.com/kopia/kopia/repo.WriteSession({0x164ef50, 0x4000fe4000}, {0xffff5277d9e0, 0x40000df4a0}, {{0x40012f4420?, 0x71d320?}, 0x10?, 0x4000442070?}, 0x400041d928)
	/go/src/github.com/kopia/kopia/repo/repository.go:361 +0x174
github.com/kopia/kopia/cli.(*App).repositoryWriterAction.func1({0x164ef50, 0x4000fe4000}, {0xffff5277d9e0, 0x40000df4a0})
	/go/src/github.com/kopia/kopia/cli/app.go:414 +0xe4
github.com/kopia/kopia/cli.(*App).maybeRepositoryAction.func1({0x164ef50, 0x4000fe4000})
	/go/src/github.com/kopia/kopia/cli/app.go:497 +0xdc
github.com/kopia/kopia/cli.(*App).baseActionWithContext.func1.1.1()
	/go/src/github.com/kopia/kopia/cli/app.go:480 +0x90
github.com/kopia/kopia/cli.(*profileFlags).withProfiling(0x0?, 0x0?)
	/go/src/github.com/kopia/kopia/cli/profile.go:45 +0x26c
github.com/kopia/kopia/cli.(*App).baseActionWithContext.func1.1({0x164ef50?, 0x40000197a0?})
	/go/src/github.com/kopia/kopia/cli/app.go:474 +0x64
github.com/kopia/kopia/cli.(*App).runAppWithContext.func1({0x164ef50, 0x40003da270}, 0x40003da270?, 0x40014e1400, 0x40002a1d50)
	/go/src/github.com/kopia/kopia/cli/app.go:445 +0x154
github.com/kopia/kopia/cli.(*App).runAppWithContext(0x40014e1400, 0xff44e9?, 0x8?)
	/go/src/github.com/kopia/kopia/cli/app.go:446 +0x110
github.com/kopia/kopia/cli.(*App).baseActionWithContext.func1(0x40002a1da8?)
	/go/src/github.com/kopia/kopia/cli/app.go:473 +0x50
github.com/alecthomas/kingpin.(*actionMixin).applyActions(0x0?, 0xfb9ce0?)
	/go/pkg/mod/github.com/alecthomas/kingpin@v0.0.0-20200323085623-b6657d9477a6/actions.go:28 +0x68
github.com/alecthomas/kingpin.(*Application).applyActions(0x40004c0500?, 0x40001743f0)
	/go/pkg/mod/github.com/alecthomas/kingpin@v0.0.0-20200323085623-b6657d9477a6/app.go:560 +0xf8
github.com/alecthomas/kingpin.(*Application).execute(0x40004c0500, 0x40001743f0?, {0x4001315300, 0x2, 0x2})
	/go/pkg/mod/github.com/alecthomas/kingpin@v0.0.0-20200323085623-b6657d9477a6/app.go:398 +0x54
github.com/alecthomas/kingpin.(*Application).Parse(0x40004c0500, {0x4001346000?, 0x4001301800?, 0x401cda488c?})
	/go/pkg/mod/github.com/alecthomas/kingpin@v0.0.0-20200323085623-b6657d9477a6/app.go:230 +0xec
github.com/kopia/kopia/cli.(*App).RunSubcommand.func2()
	/go/src/github.com/kopia/kopia/cli/inproc.go:38 +0xc8
created by github.com/kopia/kopia/cli.(*App).RunSubcommand
	/go/src/github.com/kopia/kopia/cli/inproc.go:33 +0x4cc

goroutine 2207 [select]:
io.(*pipe).read(0x40003de4e0, {0x4000420000, 0x1000, 0x5c1e4?})
	/usr/local/go/src/io/pipe.go:57 +0x84
io.(*PipeReader).Read(0x40005655f8?, {0x4000420000?, 0xe14fa0?, 0xf56340?})
	/usr/local/go/src/io/pipe.go:136 +0x28
bufio.(*Scanner).Scan(0x4000071f40)
	/usr/local/go/src/bufio/scan.go:215 +0x89c
github.com/kopia/kopia/tests/testenv.(*CLITest).Run.func1()
	/go/src/github.com/kopia/kopia/tests/testenv/cli_test_env.go:196 +0xac
created by github.com/kopia/kopia/tests/testenv.(*CLITest).Run
	/go/src/github.com/kopia/kopia/tests/testenv/cli_test_env.go:192 +0x3c4

goroutine 2324 [select]:
github.com/kopia/kopia/internal/cache.(*PersistentCache).sweepDirectoryPeriodically(0x4001346800, {0x164fea0, 0x40013347a0})
	/go/src/github.com/kopia/kopia/internal/cache/persistent_lru_cache.go:208 +0xac
created by github.com/kopia/kopia/internal/cache.NewPersistentCache
	/go/src/github.com/kopia/kopia/internal/cache/persistent_lru_cache.go:359 +0x350

goroutine 1012 [chan receive]:
github.com/kopia/kopia/cli.onCtrlC.func1()
	/go/src/github.com/kopia/kopia/cli/config.go:33 +0x30
created by github.com/kopia/kopia/cli.onCtrlC
	/go/src/github.com/kopia/kopia/cli/config.go:32 +0xc0

goroutine 226 [syscall]:
os/signal.signal_recv()
	/usr/local/go/src/runtime/sigqueue.go:151 +0x34
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:23 +0x20
created by os/signal.Notify.func1.1
	/usr/local/go/src/os/signal/signal.go:151 +0x30

goroutine 2174 [chan receive]:
github.com/kopia/kopia/cli.onCtrlC.func1()
	/go/src/github.com/kopia/kopia/cli/config.go:33 +0x30
created by github.com/kopia/kopia/cli.onCtrlC
	/go/src/github.com/kopia/kopia/cli/config.go:32 +0xc0

goroutine 1374 [chan receive]:
github.com/kopia/kopia/cli.onCtrlC.func1()
	/go/src/github.com/kopia/kopia/cli/config.go:33 +0x30
created by github.com/kopia/kopia/cli.onCtrlC
	/go/src/github.com/kopia/kopia/cli/config.go:32 +0xc0

goroutine 227 [chan receive]:
github.com/kopia/kopia/cli.onCtrlC.func1()
	/go/src/github.com/kopia/kopia/cli/config.go:33 +0x30
created by github.com/kopia/kopia/cli.onCtrlC
	/go/src/github.com/kopia/kopia/cli/config.go:32 +0xc0

goroutine 1405 [chan receive]:
github.com/kopia/kopia/cli.onCtrlC.func1()
	/go/src/github.com/kopia/kopia/cli/config.go:33 +0x30
created by github.com/kopia/kopia/cli.onCtrlC
	/go/src/github.com/kopia/kopia/cli/config.go:32 +0xc0

goroutine 2323 [select]:
github.com/kopia/kopia/internal/cache.(*PersistentCache).sweepDirectoryPeriodically(0x4001346480, {0x164fea0, 0x4001334480})
	/go/src/github.com/kopia/kopia/internal/cache/persistent_lru_cache.go:208 +0xac
created by github.com/kopia/kopia/internal/cache.NewPersistentCache
	/go/src/github.com/kopia/kopia/internal/cache/persistent_lru_cache.go:359 +0x350

goroutine 2364 [chan send]:
github.com/kopia/kopia/internal/workshare.(*AsyncGroup).RunAsync(0x40002aa8a0, 0x4000fe4f00, 0x4000f24980, {0xd36ea0?, 0x40000728d0})
	/go/src/github.com/kopia/kopia/internal/workshare/workshare_waitgroup.go:81 +0x16c
github.com/kopia/kopia/snapshot/snapshotfs.(*Uploader).processDirectoryEntries.func1({0x164ef50?, 0x40003b8b10}, {0x16545c8?, 0x4000181ef0})
	/go/src/github.com/kopia/kopia/snapshot/snapshotfs/upload.go:659 +0x2e8
github.com/kopia/kopia/fs/ignorefs.(*ignoreDirectory).IterateEntries.func1({0x164ef50, 0x40003b8b10}, {0x16545c8?, 0x4000181ef0?})
	/go/src/github.com/kopia/kopia/fs/ignorefs/ignorefs.go:163 +0x78
github.com/kopia/kopia/fs/localfs.(*filesystemDirectory).IterateEntries(0x40002afdb0, {0x164ef50, 0x40003b8b10}, 0x40002aa900)
	/go/src/github.com/kopia/kopia/fs/localfs/local_fs.go:174 +0x28c
github.com/kopia/kopia/fs/ignorefs.(*ignoreDirectory).IterateEntries(0x4000fe5380, {0x164ef50, 0x40003b8b10}, 0x40000e80e0)
	/go/src/github.com/kopia/kopia/fs/ignorefs/ignorefs.go:161 +0xec
github.com/kopia/kopia/snapshot/snapshotfs.(*Uploader).processDirectoryEntries(0x4000260b40, {0x164ef50, 0x40003b8b10}, 0x40000728b0, 0x400003e120, {0x4000481130, 0x43}, {0x40003cade0, 0xd}, {0x1656458, ...}, ...)
	/go/src/github.com/kopia/kopia/snapshot/snapshotfs/upload.go:651 +0x1a0
github.com/kopia/kopia/snapshot/snapshotfs.(*Uploader).processChildren(0x4000260b40, {0x164ef50, 0x40003b8b10}, 0x4000181ea0?, 0x4000260b40?, {0x4000481130, 0x43}, {0x40003cade0, 0xd}, {0x1656458, ...}, ...)
	/go/src/github.com/kopia/kopia/snapshot/snapshotfs/upload.go:542 +0xb0
github.com/kopia/kopia/snapshot/snapshotfs.uploadDirInternal({0x164ef50, 0x4000fe4fc0}, 0x4000260b40, {0x1656458, 0x4000fe5380}, 0x40001a8f48, {0x4000072890, 0x1, 0x1}, {0x4000481130, ...}, ...)
	/go/src/github.com/kopia/kopia/snapshot/snapshotfs/upload.go:1015 +0x77c
github.com/kopia/kopia/snapshot/snapshotfs.(*Uploader).processSingle(0x4000260b40, {0x164ef50, 0x4000fe4fc0}, {0x16544a8, 0x4000fe5380}, {0x40003cade0, 0xd}, 0x40005e1c28?, 0x400000eb58, {0x4000561de0, ...}, ...)
	/go/src/github.com/kopia/kopia/snapshot/snapshotfs/upload.go:734 +0x404
github.com/kopia/kopia/snapshot/snapshotfs.(*Uploader).processDirectoryEntries.func1.1(0x40003fbf80?, {0xd36ea0?, 0x40014fa300?})
	/go/src/github.com/kopia/kopia/snapshot/snapshotfs/upload.go:661 +0x84
github.com/kopia/kopia/internal/workshare.NewPool.func1()
	/go/src/github.com/kopia/kopia/internal/workshare/workshare_pool.go:62 +0x128
created by github.com/kopia/kopia/internal/workshare.NewPool
	/go/src/github.com/kopia/kopia/internal/workshare/workshare_pool.go:55 +0xd4

goroutine 715 [chan receive]:
github.com/kopia/kopia/cli.onCtrlC.func1()
	/go/src/github.com/kopia/kopia/cli/config.go:33 +0x30
created by github.com/kopia/kopia/cli.onCtrlC
	/go/src/github.com/kopia/kopia/cli/config.go:32 +0xc0

goroutine 2362 [chan receive]:
github.com/kopia/kopia/cli.onCtrlC.func1()
	/go/src/github.com/kopia/kopia/cli/config.go:33 +0x30
created by github.com/kopia/kopia/cli.onCtrlC
	/go/src/github.com/kopia/kopia/cli/config.go:32 +0xc0

goroutine 929 [chan receive]:
github.com/kopia/kopia/cli.onCtrlC.func1()
	/go/src/github.com/kopia/kopia/cli/config.go:33 +0x30
created by github.com/kopia/kopia/cli.onCtrlC
	/go/src/github.com/kopia/kopia/cli/config.go:32 +0xc0

goroutine 432 [chan receive]:
github.com/kopia/kopia/cli.onCtrlC.func1()
	/go/src/github.com/kopia/kopia/cli/config.go:33 +0x30
created by github.com/kopia/kopia/cli.onCtrlC
	/go/src/github.com/kopia/kopia/cli/config.go:32 +0xc0

goroutine 1211 [chan receive]:
github.com/kopia/kopia/cli.onCtrlC.func1()
	/go/src/github.com/kopia/kopia/cli/config.go:33 +0x30
created by github.com/kopia/kopia/cli.onCtrlC
	/go/src/github.com/kopia/kopia/cli/config.go:32 +0xc0

goroutine 521 [chan receive]:
github.com/kopia/kopia/cli.onCtrlC.func1()
	/go/src/github.com/kopia/kopia/cli/config.go:33 +0x30
created by github.com/kopia/kopia/cli.onCtrlC
	/go/src/github.com/kopia/kopia/cli/config.go:32 +0xc0

goroutine 1667 [chan receive]:
github.com/kopia/kopia/cli.onCtrlC.func1()
	/go/src/github.com/kopia/kopia/cli/config.go:33 +0x30
created by github.com/kopia/kopia/cli.onCtrlC
	/go/src/github.com/kopia/kopia/cli/config.go:32 +0xc0

goroutine 2325 [select]:
github.com/kopia/kopia/internal/cache.(*PersistentCache).sweepDirectoryPeriodically(0x4001346a80, {0x164fea0, 0x4001334ca0})
	/go/src/github.com/kopia/kopia/internal/cache/persistent_lru_cache.go:208 +0xac
created by github.com/kopia/kopia/internal/cache.NewPersistentCache
	/go/src/github.com/kopia/kopia/internal/cache/persistent_lru_cache.go:359 +0x350

goroutine 1978 [chan receive]:
github.com/kopia/kopia/cli.onCtrlC.func1()
	/go/src/github.com/kopia/kopia/cli/config.go:33 +0x30
created by github.com/kopia/kopia/cli.onCtrlC
	/go/src/github.com/kopia/kopia/cli/config.go:32 +0xc0

goroutine 1868 [chan receive]:
github.com/kopia/kopia/cli.onCtrlC.func1()
	/go/src/github.com/kopia/kopia/cli/config.go:33 +0x30
created by github.com/kopia/kopia/cli.onCtrlC
	/go/src/github.com/kopia/kopia/cli/config.go:32 +0xc0
FAIL	github.com/kopia/kopia/tests/end_to_end_test	60.073s
FAIL
```

--- a/internal/workshare/workshare_test.go
+++ b/internal/workshare/workshare_test.go
@@ -102,6 +102,24 @@ func TestComputeTreeSumNegative(t *testing.T) {
 	testComputeTreeSum(t, -1)
 }
 
+func TestDisallowedUse(t *testing.T) {
+	w := workshare.NewPool(1)
+
+	var ag workshare.AsyncGroup
+
+	w.Close()
+
+	require.Panics(t, func() {
+		ag.CanShareWork(w)
+	})
+
+	require.Panics(t, func() {
+		ag.RunAsync(w, func(c *workshare.Pool, request interface{}) {
+			t.Fatal("should not be called")
+		}, nil)
+	})
+}
+
 // nolint:thelper
 func testComputeTreeSum(t *testing.T, numWorkers int) {
 	w := workshare.NewPool(numWorkers)

--- a/internal/workshare/workshare_test.go
+++ b/internal/workshare/workshare_test.go
@@ -102,7 +102,27 @@ func TestComputeTreeSumNegative(t *testing.T) {
 	testComputeTreeSum(t, -1)
 }
 
-func TestDisallowedUse(t *testing.T) {
+func TestDisallowed_DoubleWait(t *testing.T) {
+	var ag workshare.AsyncGroup
+
+	ag.Wait()
+	require.Panics(t, func() {
+		ag.Wait()
+	})
+}
+
+func TestDisallowed_WaitAfterClose(t *testing.T) {
+	var ag workshare.AsyncGroup
+
+	ag.Close()
+	require.Panics(t, func() {
+		ag.Wait()
+	})
+
+	ag.Close() // no-op
+}
+
+func TestDisallowed_UseAfterPoolClose(t *testing.T) {
 	w := workshare.NewPool(1)
 
 	var ag workshare.AsyncGroup

--- a/internal/workshare/workshare_waitgroup.go
+++ b/internal/workshare/workshare_waitgroup.go
@@ -108,7 +108,14 @@ func (g *AsyncGroup) CanShareWork(w *Pool) bool {
 		// we successfully added token to the channel, because we have exactly the same number
 		// of workers as the capacity of the channel, one worker will wake up to process
 		// item from w.work, which will be added by RunAsync().
-		return true
+
+		// double check we're not closed
+		select {
+		case <-w.closed:
+			panic("invalid usage - CanShareWork() after workshare.AsyncGroup has been closed")
+		default:
+			return true
+		}
 
 	case <-w.closed:
 		panic("invalid usage - CanShareWork() after workshare.AsyncGroup has been closed")

--- a/internal/workshare/workshare_waitgroup.go
+++ b/internal/workshare/workshare_waitgroup.go
@@ -108,9 +108,8 @@ func (g *AsyncGroup) CanShareWork(w *Pool) bool {
 		// we successfully added token to the channel, because we have exactly the same number
 		// of workers as the capacity of the channel, one worker will wake up to process
 		// item from w.work, which will be added by RunAsync().
-
-		// double check we're not closed
 		select {
+		// double check we're not closed
 		case <-w.closed:
 			panic("invalid usage - CanShareWork() after workshare.AsyncGroup has been closed")
 		default:

--- a/internal/workshare/workshare_waitgroup.go
+++ b/internal/workshare/workshare_waitgroup.go
@@ -54,24 +54,41 @@ type AsyncGroup struct {
 	wg       *sync.WaitGroup
 	requests []interface{}
 	waited   bool
+	isClosed bool
 }
 
 // Wait waits for scheduled asynchronous work to complete and returns all asynchronously processed inputs.
 func (g *AsyncGroup) Wait() []interface{} {
-	if g.wg == nil || g.waited {
+	if g.isClosed {
+		panic("Wait() can't be called after Close()")
+	}
+
+	if g.waited {
+		panic("Wait() can't be called more than once")
+	}
+
+	g.waited = true
+
+	if g.wg == nil {
 		return nil
 	}
 
 	g.wg.Wait()
-
-	g.waited = true
 
 	return g.requests
 }
 
 // Close ensures all asynchronous work has been awaited for.
 func (g *AsyncGroup) Close() {
-	g.Wait()
+	if g.isClosed {
+		return
+	}
+
+	if !g.waited {
+		g.Wait() // ensure we wait
+	}
+
+	g.isClosed = true
 }
 
 // RunAsync starts the asynchronous work to process the provided request,
@@ -103,24 +120,22 @@ func (g *AsyncGroup) RunAsync(w *Pool, process ProcessFunc, request interface{})
 // If the function returns true, the use MUST call RunAsync() exactly once. This pattern avoids
 // allocations required to create asynchronous input if the worker pool is full.
 func (g *AsyncGroup) CanShareWork(w *Pool) bool {
+	// check pool is not closed
 	select {
-	case w.semaphore <- struct{}{}:
-		// we successfully added token to the channel, because we have exactly the same number
-		// of workers as the capacity of the channel, one worker will wake up to process
-		// item from w.work, which will be added by RunAsync().
-		select {
-		// double check we're not closed
-		case <-w.closed:
-			panic("invalid usage - CanShareWork() after workshare.AsyncGroup has been closed")
-		default:
-			return true
-		}
-
 	case <-w.closed:
 		panic("invalid usage - CanShareWork() after workshare.AsyncGroup has been closed")
 
 	default:
-		// all workers are busy.
-		return false
+		select {
+		case w.semaphore <- struct{}{}:
+			// we successfully added token to the channel, because we have exactly the same number
+			// of workers as the capacity of the channel, one worker will wake up to process
+			// item from w.work, which will be added by RunAsync().
+			return true
+
+		default:
+			// all workers are busy.
+			return false
+		}
 	}
 }

--- a/internal/workshare/workshare_waitgroup.go
+++ b/internal/workshare/workshare_waitgroup.go
@@ -14,6 +14,7 @@
 //
 // func visitNode(p *workshare.Pool, n *someNode) error {
 //   var wg workshare.AsyncGroup
+//   defer wg.Close()
 //
 //   for _, child := range n.children {
 //     if wg.CanShareWork(p) {
@@ -52,17 +53,25 @@ import (
 type AsyncGroup struct {
 	wg       *sync.WaitGroup
 	requests []interface{}
+	waited   bool
 }
 
 // Wait waits for scheduled asynchronous work to complete and returns all asynchronously processed inputs.
 func (g *AsyncGroup) Wait() []interface{} {
-	if g.wg == nil {
+	if g.wg == nil || g.waited {
 		return nil
 	}
 
 	g.wg.Wait()
 
+	g.waited = true
+
 	return g.requests
+}
+
+// Close ensures all asynchronous work has been awaited for.
+func (g *AsyncGroup) Close() {
+	g.Wait()
 }
 
 // RunAsync starts the asynchronous work to process the provided request,
@@ -78,10 +87,15 @@ func (g *AsyncGroup) RunAsync(w *Pool, process ProcessFunc, request interface{})
 		g.requests = append(g.requests, request)
 	}
 
-	w.work <- workItem{
+	select {
+	case w.work <- workItem{
 		process: process,
 		request: request,
 		wg:      g.wg,
+	}:
+
+	case <-w.closed:
+		panic("invalid usage - RunAsync() after workshare.AsyncGroup has been closed")
 	}
 }
 
@@ -95,6 +109,9 @@ func (g *AsyncGroup) CanShareWork(w *Pool) bool {
 		// of workers as the capacity of the channel, one worker will wake up to process
 		// item from w.work, which will be added by RunAsync().
 		return true
+
+	case <-w.closed:
+		panic("invalid usage - CanShareWork() after workshare.AsyncGroup has been closed")
 
 	default:
 		// all workers are busy.

--- a/snapshot/snapshotfs/dir_rewriter.go
+++ b/snapshot/snapshotfs/dir_rewriter.go
@@ -143,6 +143,9 @@ func (rw *DirRewriter) processDirectoryEntries(ctx context.Context, parentPath s
 		wg      workshare.AsyncGroup
 	)
 
+	// ensure we wait for all work items before returning
+	defer wg.Close()
+
 	for _, child := range entries {
 		if wg.CanShareWork(rw.ws) {
 			// see if we can run this child in a goroutine

--- a/snapshot/snapshotfs/snapshot_tree_walker.go
+++ b/snapshot/snapshotfs/snapshot_tree_walker.go
@@ -110,7 +110,7 @@ func (w *TreeWalker) processDirEntry(ctx context.Context, dir fs.Directory, entr
 	}
 
 	var ag workshare.AsyncGroup
-	defer ag.Wait()
+	defer ag.Close()
 
 	err := dir.IterateEntries(ctx, func(c context.Context, ent fs.Entry) error {
 		if w.TooManyErrors() {

--- a/snapshot/snapshotfs/upload.go
+++ b/snapshot/snapshotfs/upload.go
@@ -538,6 +538,9 @@ func (u *Uploader) processChildren(
 ) error {
 	var wg workshare.AsyncGroup
 
+	// ensure we wait for all work items before returning
+	defer wg.Close()
+
 	// ignore errCancel because a more serious error may be reported in wg.Wait()
 	// we'll check for cancelation later.
 

